pkg/operator/encryption/shared_encryption.go

```diff
@@ -0,0 +1,431 @@
+	return revision, nil
+}
+
+func isPodReady(pod corev1.Pod) bool {
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func getEncryptionConfig(secrets corev1client.SecretInterface, revision string) (*apiserverconfigv1.EncryptionConfiguration, error) {
+	encryptionConfigSecret, err := secrets.Get(encryptionConfSecret+"-"+revision, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	encryptionConfigObj, err := runtime.Decode(decoder, encryptionConfigSecret.Data[encryptionConfSecret])
+	if err != nil {
+		return nil, err
+	}
+
+	encryptionConfig, ok := encryptionConfigObj.(*apiserverconfigv1.EncryptionConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("encryption config has wrong type %T", encryptionConfigObj)
+	}
+	return encryptionConfig, nil
+}
+
+type actualKeys struct {
+	hasWriteKey bool
+	writeKey    apiserverconfigv1.Key
+	readKeys    []apiserverconfigv1.Key
+}
+
+func getGRsActualKeys(encryptionConfig *apiserverconfigv1.EncryptionConfiguration) map[schema.GroupResource]actualKeys {
+	out := map[schema.GroupResource]actualKeys{}
+	for _, resourceConfig := range encryptionConfig.Resources {
+		if len(resourceConfig.Resources) == 0 || len(resourceConfig.Providers) < 2 {
+			continue // should never happen
+		}
+
+		gr := schema.ParseGroupResource(resourceConfig.Resources[0])
+		provider1 := resourceConfig.Providers[0]
+		provider2 := resourceConfig.Providers[1]
+
+		switch {
+		case provider1.AESCBC != nil && len(provider1.AESCBC.Keys) != 0:
+			out[gr] = actualKeys{
+				hasWriteKey: true,
+				writeKey:    provider1.AESCBC.Keys[0],
+				readKeys:    provider1.AESCBC.Keys[1:],
+			}
+		case provider1.Identity != nil && provider2.AESCBC != nil && len(provider2.AESCBC.Keys) != 0:
+			out[gr] = actualKeys{
+				readKeys: provider2.AESCBC.Keys,
+			}
+		}
+	}
+	return out
+}
+
+func setSecretAnnotation(secretClient corev1client.SecretsGetter, recorder events.Recorder, secret *corev1.Secret, annotation string) error {
+	if len(secret.Annotations[annotation]) != 0 {
+		return nil
+	}
+	secret = secret.DeepCopy()
+
+	if secret.Annotations == nil {
+		secret.Annotations = map[string]string{}
+	}
+	secret.Annotations[annotation] = time.Now().Format(time.RFC3339)
+
+	_, _, updateErr := resourceapply.ApplySecret(secretClient, recorder, secret)
+	return updateErr // let conflict errors cause a retry
+}
+
+func setUpGlobalMachineConfigEncryptionInformers(
+	operatorClient operatorv1helpers.StaticPodOperatorClient,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	eventHandler cache.ResourceEventHandler,
+) []cache.InformerSynced {
+	operatorInformer := operatorClient.Informer()
+	operatorInformer.AddEventHandler(eventHandler)
+
+	secretsInformer := kubeInformersForNamespaces.InformersFor(operatorclient.GlobalMachineSpecifiedConfigNamespace).Core().V1().Secrets().Informer()
```

Comment:
I asked myself whether it is safe to assume that an informer will be created for that namespace. I think it is because we do `secretClient.Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace)` which uses `CachedSecretGetter`.